src/mlpack/methods/ann/layer/adaptive_mean_pooling.hpp

```diff
@@ -0,0 +1,267 @@
+   */
+  AdaptiveMeanPooling(const std::tuple<size_t, size_t> outputShape);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, using 3rd-order tensors as
+   * input, calculating the function f(x) by propagating x backwards through f.
+   * Using the results from the feed forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& /* input */,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  //! Get the output parameter.
+  OutputDataType const &OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType &OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const &Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType &Delta() { return delta; }
+
+  //! Get the width.
+  size_t const &InputWidth() const { return inputWidth; }
+  //! Modify the width.
+  size_t &InputWidth() { return inputWidth; }
+
+  //! Get the height.
+  size_t const &InputHeight() const { return inputHeight; }
+  //! Modify the height.
+  size_t &InputHeight() { return inputHeight; }
+
+  //! Get the width.
+  size_t const &OutputWidth() const { return outputWidth; }
+  //! Modify the width.
+  size_t &OutputWidth() { return outputWidth; }
+
+  //! Get the height.
+  size_t const &OutputHeight() const { return outputHeight; }
+  //! Modify the height.
+  size_t &OutputHeight() { return outputHeight; }
+
+  //! Get the input size.
+  size_t InputSize() const { return inSize; }
+
+  //! Get the output size.
+  size_t OutputSize() const { return outSize; }
+
+  //! Get the value of the deterministic parameter.
+  bool Deterministic() const { return deterministic; }
+  //! Modify the value of the deterministic parameter.
+  bool &Deterministic() { return deterministic; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  /**
+   * Initialize Kernel Size and Stride for Adaptive Pooling.
+   */
+  void IntializeAdaptivePadding()
+  {
+    strideWidth = std::floor(inputWidth / outputWidth);
+    strideHeight = std::floor(inputHeight / outputHeight);
+
+    kernelWidth = inputWidth - (outputWidth - 1) * strideWidth;
+    kernelHeight = inputHeight - (outputHeight - 1) * strideHeight;
+
+    if (kernelHeight < 0 || kernelWidth < 0)
+    {
+      Log::Fatal << "Given output shape is not possible for given "
```

Comment:
Might be a good idea, to print the input/output shape as well. What do you think?

Reply:
Agreed, that's a better idea.